File: rust/project/fielded.rb
module RustProjection
  module Projector
    module_function

    # ── FIELDED — one match arm per real attribute, generated identically
    # for value objects, command args structs, and (via emit_fielded_record,
    # below) aggregate records. This is the mechanical glue Rust's lack of
    # reflection still requires — READING a named field generically, unlike
    # WRITING one, needs no per-command bespoke control flow, only this
    # uniform per-type table.
    # `extra_arms:` — raw `"key" => ...,` lines appended verbatim, never
    # run through the attribute-typed branches above. Exists for the
    # LIFECYCLE field on an ENTITY: `emit_entity` gives an entity struct a
    # bare `String` field for it (the same as `emit_record` does for an
    # aggregate), but unlike `emit_record` (which routes through
    # `emit_fielded_record`, its own record-shaped sibling with a
    # lifecycle arm built in), `emit_fielded_flat` had no lifecycle
    # awareness at all until an entity command's own `TransitionCheck`
    # needed to read it generically — the same gap `emit_to_json_flat`'s
    # `extra_fields:` closes for JSON, here for `Fielded` lookup.
    #
    # Each arm below renders from one of six leaf shapes in
    # rust/src/exemplar/fielded.rs (`fielded_arm_list_optional`,
    # `_list`, `_optional_scalar`, `_optional_nested`, `_scalar`,
    # `_nested`) — flush-left text, the SAME fixed 12-space prefix the
    # original hand-built strings baked in added back here, not through
    # `compose`'s auto-reindent (see `closed_set_table_codec`'s own
    # header, json_codec.rb, for why plain `render` is the right choice
    # when a leaf is reused across more than one outer shape).
    def emit_fielded_flat(struct_name, attributes, value_objects_by_name, extra_arms: [], entity_names: [])
      arms = attributes.filter_map do |attr|
        key   = rust_field(attr[:name])
        ident = rust_ident_field(attr[:name])
        scalar = effective_scalar_type(attr[:type])
        if attr[:list] && attr[:optional]
          Exemplar.render("fielded_arm_list_optional", '"tmpl_field"' => key.inspect, "tmpl_ident" => ident)
        elsif attr[:list]
          Exemplar.render("fielded_arm_list", '"tmpl_field"' => key.inspect, "tmpl_ident" => ident)
        elsif attr[:optional] && scalar
          Exemplar.render(
            "fielded_arm_optional_scalar",
            '"tmpl_field"' => key.inspect,
            "tmpl_ident" => ident,
            "tmpl_value_expr_placeholder(v)" => scalar_to_value(scalar, "v")
          )
        elsif attr[:optional]
          nested = value_objects_by_name[attr[:type]]
          next nil unless nested && fielded_capable_nested?(nested)

          Exemplar.render("fielded_arm_optional_nested", '"tmpl_field"' => key.inspect, "tmpl_ident" => ident)
        elsif scalar
          Exemplar.render(
            "fielded_arm_scalar",
            '"tmpl_field"' => key.inspect,
            "tmpl_value_expr_placeholder(&self.tmpl_ident)" => scalar_to_value(scalar, "self.#{ident}")
          )
        else
          nested = value_objects_by_name[attr[:type]]
          next nil unless nested && fielded_capable_nested?(nested)

          Exemplar.render("fielded_arm_nested", '"tmpl_field"' => key.inspect, "tmpl_ident" => ident)
        end
      end
      arms = arms.map { |a| "            #{a}" }
      arms += extra_arms

      # The trailing "\n" restores the OLD heredoc's own implicit one
      # (`<<~RUST` always ends its result in a newline) — every caller's
      # own blank-line arithmetic between this and whatever comes next
      # (`emit_value_object`, `emit_entity`) was written assuming it,
      # and `Exemplar.render`'s dedent `.rstrip`s trailing whitespace,
      # which would otherwise silently collapse two blank lines to one.
      "#{Exemplar.render(
        'fielded_flat',
        'TmplFlatType' => struct_name,
        '"tmpl_arms_placeholder" => tmpl_arms_block(),' => arms.join("\n"),
        # `entity_names` — a value object that holds a LIST OF ENTITIES
        # (chess's `Position`, one snapshot of the board's own piece
        # lists) enumerates them the same way the record does.
        '"tmpl_items_placeholder" => tmpl_items_block(),' => items_arms(attributes, value_objects_by_name, entity_names, optional: ->(attr) { attr[:optional] }).join("\n"),
        "tmpl_as_scalar_placeholder()" => as_scalar_expr(attributes),
        # Same conditional-import trick as `Value` just below — a struct
        # with no fielded-capable attributes (a bare command like
        # `Retire`, whose `attributes` is empty and `extra_arms` stays
        # empty too) renders `match name { _ => None, }`, never
        # constructing a bare `Field::...` — leaving `use
        # crate::kernel::Field;` unused. Found live as a real `cargo
        # check --workspace` warning across several generated files
        # (RoleAssignment, Widget's `RetireArgs`, ...).
        'use crate::kernel::Field;' => (arms.any? { |arm| arm.include?('Field') } ? 'use crate::kernel::Field;' : ''),
        'use crate::kernel::Value;' => (arms.any? { |arm| arm.include?('Value') } ? 'use crate::kernel::Value;' : '')
      )}\n"
    end

    # `Fielded::items` — one arm per list attribute whose ELEMENT this
    # generator can hand the kernel: a scalar (through `scalar_to_value`,
    # the same reading the scalar field arms use) or a nested value
    # object/entity that has a `Fielded` impl of its own. A list of
    # anything else gets no arm and stays a length-only field, exactly
    # as before the enumeration operators existed. `optional` decides
    # per attribute whether the list is `Option`-wrapped — the flat and
    # record emitters answer that differently (see each caller).
    def items_arms(attributes, value_objects_by_name, entity_names, optional:)
      attributes.filter_map do |attr|
        next nil unless attr[:list]

        key    = rust_field(attr[:name])
        ident  = rust_ident_field(attr[:name])
        scalar = effective_scalar_type(attr[:type])
        nested = value_objects_by_name[attr[:type]]
        fielded_element = scalar || (nested && fielded_capable_nested?(nested)) || entity_names.include?(attr[:type])
        next nil unless fielded_element

        shape = "fielded_items_arm_list_#{optional.call(attr) ? 'optional_' : ''}#{scalar ? 'scalar' : 'nested'}"
        subs = { '"tmpl_field"' => key.inspect, "tmpl_ident" => ident }
        subs["tmpl_value_expr_placeholder(v)"] = scalar_to_value(scalar, "v") if scalar
        "            #{Exemplar.render(shape, subs)}"
      end
    end

    # Aggregate records differ from value objects/args structs in one main
    # way — every non-list attribute is `Option`-wrapped (see emit_record's
    # own comment for why), so a `None` field reads as `Value::Nil` rather
    # than being absent from the match at all — PLUS the one list-typed
    # exception `emit_record` itself now makes (`list_attr_creation_
    # optional?`, mutations.rb). Its `scalar`/`nested` arms reuse
    # `emit_fielded_flat`'s own `_optional_scalar`/`_optional_nested`
    # shapes verbatim (blanket-Option-wrapped, same as those two already
    # are) — only `list`/`list_optional` and the lifecycle arm are
    # rendered here directly.
    def emit_fielded_record(aggregate, value_objects_by_name)
      name = rust_ident(aggregate[:name])
      arms = aggregate[:attributes].filter_map do |attr|
        key   = rust_field(attr[:name])
        ident = rust_ident_field(attr[:name])
        scalar = effective_scalar_type(attr[:type])
        if attr[:list] && list_attr_creation_optional?(aggregate, attr[:name], value_objects_by_name)
          Exemplar.render("fielded_arm_list_optional", '"tmpl_field"' => key.inspect, "tmpl_ident" => ident)
        elsif attr[:list]
          Exemplar.render("fielded_arm_list", '"tmpl_field"' => key.inspect, "tmpl_ident" => ident)
        elsif scalar
          Exemplar.render(
            "fielded_arm_optional_scalar",
            '"tmpl_field"' => key.inspect,
            "tmpl_ident" => ident,
            "tmpl_value_expr_placeholder(v)" => scalar_to_value(scalar, "v")
          )
        else
          nested = value_objects_by_name[attr[:type]]
          next nil unless nested && fielded_capable_nested?(nested)

          Exemplar.render("fielded_arm_optional_nested", '"tmpl_field"' => key.inspect, "tmpl_ident" => ident)
        end
      end
      if aggregate[:lifecycle]
        key   = rust_field(aggregate[:lifecycle][:field])
        ident = rust_ident_field(aggregate[:lifecycle][:field])
        arms << Exemplar.render("fielded_lifecycle_arm", '"tmpl_field"' => key.inspect, "tmpl_ident" => ident)
      end
      correctable_event_names(aggregate).each do |ev|
        ident = corrects_flag_field(ev)
        arms << Exemplar.render("fielded_corrects_flag_arm", '"tmpl_field"' => ident.inspect, "tmpl_ident" => ident)
      end
      arms = arms.map { |a| "            #{a}" }

      # Trailing "\n" — see `emit_fielded_flat`'s own comment on why.
      entity_names = (aggregate[:entities] || []).map { |e| e[:name] }
      "#{Exemplar.render(
        'fielded_record',
        'TmplRecordType' => name,
        '"tmpl_arms_placeholder" => tmpl_arms_block(),' => arms.join("\n"),
        '"tmpl_items_placeholder" => tmpl_items_block(),' => items_arms(
          aggregate[:attributes], value_objects_by_name, entity_names,
          optional: ->(attr) { list_attr_creation_optional?(aggregate, attr[:name], value_objects_by_name) }
        ).join("\n"),
        "tmpl_as_scalar_placeholder()" => as_scalar_expr(aggregate[:attributes])
      )}\n"
    end

    # `Resolver#unwrap_scalar` — a struct with exactly ONE attribute
    # reads as that attribute's value, WHATEVER it is named (see the
    # exemplar's own `tmpl_as_scalar_placeholder` comment); anything
    # else answers `None`. Used to gate on the sole attribute being
    # literally NAMED `value` — the same name-vs-count split the Ruby
    # oracle's own `unwrap_scalar` carried, relaxed in lockstep with it
    # (single-element value objects strictly answer `.value`,
    # [[feedback_name_the_scalar_field]]): a `Money{amount}` IS its
    # amount exactly as a `Label{value}` is its value, and the name gate
    # made the two behave differently in every generated predicate walk.
    # Only a genuine SCALAR leaf unwraps — a sole attribute that is
    # itself a value object (`Field::Nested`, or a closed set with no
    # arm at all) already answered `None` through the match's own `_`
    # floor, so requiring `effective_scalar_type` here changes no
    # runtime answer; it just emits the honest literal `None` instead of
    # a match that could never bind (and stops the generated text
    # mentioning a field name no arm exists for — spec/rust_project/
    # closed_set_fielded_spec.rb pins exactly that for the multi-field
    # closed-set shape).
    def as_scalar_expr(attributes)
      sole = attributes.size == 1 && !attributes.first[:list] ? attributes.first : nil
      return "None" unless sole && effective_scalar_type(sole[:type])

      "match self.field(#{rust_field(sole[:name]).inspect}) { Some(crate::kernel::Field::Value(v)) => Some(v), _ => None }"
    end
  end
end

File: rust/codegen/src/fielded.rs
//! Port of `rust/project/fielded.rb` — `emit_fielded_flat`/
//! `emit_fielded_record`, mirrored directly. Read the Ruby file's own
//! header for the six/seven leaf shapes this walks.

use crate::exemplar::Exemplar;
use crate::json::Json;
use crate::naming;
use std::collections::HashMap;

fn is_fielded_nested(attr_type: &str, value_objects_by_name: &HashMap<String, &Json>) -> bool {
    match value_objects_by_name.get(attr_type) {
        Some(vo) => naming::fielded_capable_nested(vo),
        None => false,
    }
}

pub fn emit_fielded_flat(exemplar: &Exemplar, struct_name: &str, attributes: &[Json], value_objects_by_name: &HashMap<String, &Json>, extra_arms: &[String]) -> String {
    let mut arms: Vec<String> = Vec::new();

    for attr in attributes {
        let key = naming::rust_field(crate::attr::name(attr));
        let ident = naming::rust_ident_field(crate::attr::name(attr));
        let scalar = naming::effective_scalar_type(crate::attr::type_name(attr));
        let list = crate::attr::list(attr);
        let optional = crate::attr::optional(attr);

        let arm = if list && optional {
            Some(exemplar.render("fielded_arm_list_optional", &[("\"tmpl_field\"", naming::ruby_inspect_string(&key)), ("tmpl_ident", ident)]))
        } else if list {
            Some(exemplar.render("fielded_arm_list", &[("\"tmpl_field\"", naming::ruby_inspect_string(&key)), ("tmpl_ident", ident)]))
        } else if optional && scalar.is_some() {
            let scalar = scalar.unwrap();
            Some(exemplar.render(
                "fielded_arm_optional_scalar",
                &[("\"tmpl_field\"", naming::ruby_inspect_string(&key)), ("tmpl_ident", ident.clone()), ("tmpl_value_expr_placeholder(v)", naming::scalar_to_value(scalar, "v").unwrap())],
            ))
        } else if optional {
            if is_fielded_nested(crate::attr::type_name(attr), value_objects_by_name) {
                Some(exemplar.render("fielded_arm_optional_nested", &[("\"tmpl_field\"", naming::ruby_inspect_string(&key)), ("tmpl_ident", ident)]))
            } else {
                None
            }
        } else if let Some(scalar) = scalar {
            Some(exemplar.render(
                "fielded_arm_scalar",
                &[("\"tmpl_field\"", naming::ruby_inspect_string(&key)), ("tmpl_value_expr_placeholder(&self.tmpl_ident)", naming::scalar_to_value(scalar, &format!("self.{ident}")).unwrap())],
            ))
        } else if is_fielded_nested(crate::attr::type_name(attr), value_objects_by_name) {
            Some(exemplar.render("fielded_arm_nested", &[("\"tmpl_field\"", naming::ruby_inspect_string(&key)), ("tmpl_ident", ident)]))
        } else {
            None
        };

        if let Some(a) = arm {
            arms.push(a);
        }
    }

    let mut arms: Vec<String> = arms.iter().map(|a| format!("            {a}")).collect();
    arms.extend(extra_arms.iter().cloned());
    // `arms.any? { |arm| arm.include?('Value') }` — checked against the
    // full, already-assembled arm list (declared arms + extra_arms
    // together), not tracked incrementally per branch while building —
    // found live, byte-diffing against Ruby's real output: a plain
    // scalar arm (`fielded_arm_scalar`) also emits `Value::...` text, and
    // an incremental per-branch tracker that only flagged the list/
    // optional branches (this function's own first draft) missed it.
    let uses_value = arms.iter().any(|a| a.contains("Value"));
    // Mirrors `fielded.rb#emit_fielded_flat`'s same conditional-import
    // trick, now also applied to `Field` — a struct with no
    // fielded-capable attributes (a bare command like `Retire`) renders
    // `match name { _ => None, }`, never constructing a bare
    // `Field::...`, leaving `use crate::kernel::Field;` unused.
    let uses_field = arms.iter().any(|a| a.contains("Field"));

    format!(
        "{}\n",
        exemplar.render(
            "fielded_flat",
            &[
                ("TmplFlatType", struct_name.to_string()),
                ("\"tmpl_arms_placeholder\" => tmpl_arms_block(),", arms.join("\n")),
                ("\"tmpl_items_placeholder\" => tmpl_items_block(),", items_arms(exemplar, attributes, value_objects_by_name, &[], |attr| crate::attr::optional(attr)).join("\n")),
                ("tmpl_as_scalar_placeholder()", as_scalar_expr(attributes)),
                ("use crate::kernel::Field;", if uses_field { "use crate::kernel::Field;".to_string() } else { String::new() }),
                ("use crate::kernel::Value;", if uses_value { "use crate::kernel::Value;".to_string() } else { String::new() }),
            ],
        )
    )
}

pub fn emit_fielded_record(exemplar: &Exemplar, aggregate: &Json, value_objects_by_name: &HashMap<String, &Json>) -> String {
    let name = naming::rust_ident(aggregate.get("name").and_then(Json::as_str).unwrap_or(""));
    let attributes = aggregate.get("attributes").map(Json::each).unwrap_or(&[]);
    let mut arms: Vec<String> = Vec::new();

    for attr in attributes {
        let key = naming::rust_field(crate::attr::name(attr));
        let ident = naming::rust_ident_field(crate::attr::name(attr));
        let scalar = naming::effective_scalar_type(crate::attr::type_name(attr));
        let list = crate::attr::list(attr);

        let arm = if list && crate::shared::list_attr_creation_optional(aggregate, crate::attr::name(attr), value_objects_by_name) {
            Some(exemplar.render("fielded_arm_list_optional", &[("\"tmpl_field\"", naming::ruby_inspect_string(&key)), ("tmpl_ident", ident)]))
        } else if list {
            Some(exemplar.render("fielded_arm_list", &[("\"tmpl_field\"", naming::ruby_inspect_string(&key)), ("tmpl_ident", ident)]))
        } else if let Some(scalar) = scalar {
            Some(exemplar.render(
                "fielded_arm_optional_scalar",
                &[("\"tmpl_field\"", naming::ruby_inspect_string(&key)), ("tmpl_ident", ident.clone()), ("tmpl_value_expr_placeholder(v)", naming::scalar_to_value(scalar, "v").unwrap())],
            ))
        } else if is_fielded_nested(crate::attr::type_name(attr), value_objects_by_name) {
            Some(exemplar.render("fielded_arm_optional_nested", &[("\"tmpl_field\"", naming::ruby_inspect_string(&key)), ("tmpl_ident", ident)]))
        } else {
            None
        };

        if let Some(a) = arm {
            arms.push(a);
        }
    }

    if let Some(lifecycle) = aggregate.get("lifecycle") {
        let field = lifecycle.get("field").and_then(Json::as_str).unwrap_or("");
        let key = naming::rust_field(field);
        let ident = naming::rust_ident_field(field);
        arms.push(exemplar.render("fielded_lifecycle_arm", &[("\"tmpl_field\"", naming::ruby_inspect_string(&key)), ("tmpl_ident", ident)]));
    }
    for ev in crate::bridging::correctable_event_names(aggregate) {
        let ident = crate::bridging::corrects_flag_field(&ev);
        arms.push(exemplar.render("fielded_corrects_flag_arm", &[("\"tmpl_field\"", naming::ruby_inspect_string(&ident)), ("tmpl_ident", ident)]));
    }

    let arms: Vec<String> = arms.iter().map(|a| format!("            {a}")).collect();

    let entity_names: Vec<String> = aggregate
        .get("entities")
        .map(Json::each)
        .unwrap_or(&[])
        .iter()
        .filter_map(|e| e.get("name").and_then(Json::as_str).map(str::to_string))
        .collect();
    let items = items_arms(exemplar, attributes, value_objects_by_name, &entity_names, |attr| {
        crate::shared::list_attr_creation_optional(aggregate, crate::attr::name(attr), value_objects_by_name)
    });

    format!(
        "{}\n",
        exemplar.render(
            "fielded_record",
            &[
                ("TmplRecordType", name),
                ("\"tmpl_arms_placeholder\" => tmpl_arms_block(),", arms.join("\n")),
                ("\"tmpl_items_placeholder\" => tmpl_items_block(),", items.join("\n")),
                ("tmpl_as_scalar_placeholder()", as_scalar_expr(attributes)),
            ],
        )
    )
}

/// Port of fielded.rb's `as_scalar_expr` — `Resolver#unwrap_scalar`: a
/// struct with exactly one attribute reads as that attribute's value,
/// whatever it is named (single-element value objects strictly answer
/// `.value` — relaxed from the old name-gated `== "value"` check in
/// lockstep with the Ruby projector and the Ruby oracle's own
/// `unwrap_scalar`; see fielded.rb's `as_scalar_expr` for the full
/// account). Only a genuine scalar leaf unwraps — a sole attribute
/// that is itself a value object already answered `None` through the
/// match's own `_` floor, so gating on `effective_scalar_type` changes
/// no runtime answer; it emits the honest literal `None` instead of a
/// match that could never bind (fielded.rb's own `as_scalar_expr`
/// comment, and spec/rust_project/closed_set_fielded_spec.rb's pin).
fn as_scalar_expr(attributes: &[Json]) -> String {
    let sole = attributes.len() == 1
        && !crate::attr::list(&attributes[0])
        && naming::effective_scalar_type(crate::attr::type_name(&attributes[0])).is_some();
    if sole {
        format!(
            "match self.field(\"{}\") {{ Some(crate::kernel::Field::Value(v)) => Some(v), _ => None }}",
            naming::rust_field(crate::attr::name(&attributes[0]))
        )
    } else {
        "None".to_string()
    }
}

/// Port of `fielded.rb`'s `items_arms` — `Fielded::items`, one arm per
/// list attribute whose element is a scalar or a `Fielded` value object/
/// entity; anything else stays a length-only field with no arm.
fn items_arms(exemplar: &Exemplar, attributes: &[Json], value_objects_by_name: &HashMap<String, &Json>, entity_names: &[String], optional: impl Fn(&Json) -> bool) -> Vec<String> {
    let mut arms = Vec::new();
    for attr in attributes {
        if !crate::attr::list(attr) {
            continue;
        }
        let key = naming::rust_field(crate::attr::name(attr));
        let ident = naming::rust_ident_field(crate::attr::name(attr));
        let type_name = crate::attr::type_name(attr);
        let scalar = naming::effective_scalar_type(type_name);
        let fielded_element = scalar.is_some() || is_fielded_nested(type_name, value_objects_by_name) || entity_names.iter().any(|e| e == type_name);
        if !fielded_element {
            continue;
        }
        let shape = format!("fielded_items_arm_list_{}{}", if optional(attr) { "optional_" } else { "" }, if scalar.is_some() { "scalar" } else { "nested" });
        let mut subs: Vec<(&str, String)> = vec![("\"tmpl_field\"", naming::ruby_inspect_string(&key)), ("tmpl_ident", ident)];
        if let Some(scalar) = scalar {
            subs.push(("tmpl_value_expr_placeholder(v)", naming::scalar_to_value(scalar, "v").unwrap()));
        }
        arms.push(format!("            {}", exemplar.render(&shape, &subs)));
    }
    arms
}
